### ISSUE TITLE
Global configuration file

### ISSUE THREAD

**Issue by muuvmuuv** [Unibeautify/unibeautify-cli#104]
The VS Code extension already got the option to pass a default config file, if no config file is provided in the project itself. We could do the same in two ways:

1. we have the Unibeautify config file in the user root `~/.unibeautifyrc.json`
2. we use the `~/.unibeautifrc.json` file to set CLI options e.g.:

```json
{
  "defaultConfig": "/Users/marvinheilemann/Documents/Configs/.unibeautifyrc.json"
}
```

**Comment by lassik** [comment]
I like this. Personal default settings for quick scripts.

We could follow the conventions of [Unix desktop environments](https://standards.freedesktop.org/basedir-spec/basedir-spec-latest.html) (GNOME, KDE) and put the files in the `~/.config` directory. People have so many dotfiles in the home directory already :) The `~/.config` directory is already used by lots of programs, even command line programs.

So the full filename could be `~/.config/unibeautifyrc.json` - or `~/.config/unibeautify/rc.json` if we want Unibeautify to have its own subdirectory.

(To be pedantic,  the location of the config directory comes from the environment variable `$XDG_CONFIG_HOME` but I think practically everyone just uses the default location, `~/.config`).

**Comment by lassik** [comment]
(There's a similar convention of using the `~/.cache` directory if we ever need to cache something.)

**Comment by Glavin001** [comment]
I am **not** in support of the Unibeautify configuration file having an option pointing to an absolute path, such as in @muuvmuuv 's example #2: https://github.com/Unibeautify/unibeautify-cli/issues/104#issue-373387521

Here's the discussion we had over in VSCode project: https://github.com/Unibeautify/vscode/issues/55#issuecomment-398851249
I was OK with adding `defaultConfig` to VSCode's configuration because I see a valid workflow: having VSCode User Settings with `unibeautify.defaultConfig` pointing to say `~/.unibeautifyrc`.

I'd recommend using `unibeautify --config-file /Users/marvinheilemann/Documents/Configs/.unibeautifyrc.json`.

I think Cosmiconfig -- which Unibeautify uses -- already searches the home directory: https://github.com/davidtheclark/cosmiconfig#explorersearch
Using `search`: https://github.com/Unibeautify/unibeautify-cli/blob/74ee41fb76c154b107f29b22457100a05ece993f/src/commands/BeautifyCommand.ts#L77

**Comment by lassik** [comment]
> I am **not** in support of the Unibeautify configuration file having an option pointing to an absolute path

Agreed. I think it would be a good idea to support user defaults, but the pathname should be implicit (always in a standard location) or given to the CLI as `--config-file`, not specified in other `.unibeutifyrc` files.

**Comment by muuvmuuv** [comment]
@lassik yep, that is what I meant by 1. So the user places a config file in eg `~/.config/.unibeautifyrc.json` and Unibeautify will use this if 1. no config is defined via flag and 2. cosmiconfig search couldn‘t find a config (or maybe 3. no .editorconfig is in place)

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by Glavin001** [comment]
> So the user places a config file in eg `~/.config/.unibeautifyrc.json`

The standard path would be `~/.unibeautifyrc.json`, not `~/.config/...`.

See https://github.com/davidtheclark/cosmiconfig#explorersearch and https://github.com/davidtheclark/cosmiconfig#stopdir 

However, this may only work if your code is within your home directory.

Looks like it would not currently work because of the `stopDir` in https://github.com/Unibeautify/unibeautify-cli/blob/74ee41fb76c154b107f29b22457100a05ece993f/src/commands/BeautifyCommand.ts#L74-L77

```javascript
cosmiconfig("unibeautify", { stopDir: filePath })
```

@muuvmuuv  As mentioned in https://github.com/Unibeautify/unibeautify-cli/issues/104#issuecomment-433632111 my recommendation is to use the already supported `--config-file` CLI argument:

```
unibeautify --config-file /Users/marvinheilemann/Documents/Configs/.unibeautifyrc.json
```

---

My problem with allowing these global configuration files is they should really be stored within the project itself so they can be shared throughout your team. It means if I clone your repository, I am not able to reproduce the same results and therefore creates a confusing experience.

At this time, I'm not confident I would even accept a Pull Request implementing this global configuration support by changing `stopDir` 🤔 .

**Comment by lassik** [comment]
> My problem with allowing these global configuration files is they should really be stored within the project itself so they can be shared throughout your team. It means if I clone your repository, I am not able to reproduce the same results and therefore creates a confusing experience.

I fully agree, but the point of the global file would be to serve defaults for one-off hacks (e.g. when you spend an hour writing a throwaway script - code formatting is still very useful in these situations). It could also serve as the default for temporary code in a text editor that isn't saved to any file (I often do this several times a day to work out some ideas). Similar to the way text editors have default settings for coding style but can change them per project.

> > So the user places a config file in eg 
> 
> The standard path would be `~/.unibeautifyrc.json`, not `~/.config/...`.

It seems all of us are actually confusing two different approaches here.

* One is to look up in parent directories until reaching the home directory (what CosmicConfig seems to do).

* The other approach is to look right away in a known default location (`~/.unibeautifyrc.json` or `~/.config/.unibeautifyrc.json`) no matter where the file to beautify is (even if it's outside the user's home directory).

It's true that traditionally config files have gone directly under the homedir, but `~/.config` has also been a de facto standard for years. Almost everyone has that directory already because lots of programs put their stuff in there.

**Comment by lassik** [comment]
Another way to put it: to *inherit/extend* a coding style should be a separate matter from the *default* coding style. The default style should only be used for project that don't have any config file of their own.

**Comment by lassik** [comment]
Good thinking. In light of this, I would suggest one of the following:

* Look for the default config file if there is no project config file only when `--default` is given.
* Look for the default config file if there is no project config file unless `--nodefault` is given.

The right spelling (`default` vs `defaults`,  `nodefault` vs `no-default`) is not clear.
